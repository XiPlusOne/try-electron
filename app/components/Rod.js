import React from 'react';
import styles from './Rod.less';

type Props = {
  painter: {
    paint: HTMLCanvasElement => void
  },
  style?: Object,
  isRodOn?: boolean,
  rodOnCallback?: () => void,
  rodOffCallback?: () => void
};

export default class Rod extends React.PureComponent {
  props: Props;

  static defaultProps = {
    style: null,
    isRodOn: false,
    rodOnCallback: null,
    rodOffCallback: null
  };

  constructor(props) {
    super(props);

    this.refCanvas = React.createRef();
    this.onTransitionEnd = this.onTransitionEnd.bind(this);
  }

  componentDidMount() {
    const { painter } = this.props;

    painter.paint(this.refCanvas.current);
  }

  onTransitionEnd() {
    const { isRodOn, rodOnCallback, rodOffCallback } = this.props;

    if (typeof rodOnCallback === 'function' && isRodOn) {
      rodOnCallback();
    }

    if (typeof rodOffCallback === 'function' && !isRodOn) {
      rodOffCallback();
    }
  }

  render() {
    const { isRodOn, style } = this.props;

    return (
      <div className={styles.probe} style={style}>
        <b />
        <canvas
          ref={this.refCanvas}
          className={isRodOn ? styles.on : styles.off}
          onTransitionEnd={this.onTransitionEnd}
        />
      </div>
    );
  }
}
