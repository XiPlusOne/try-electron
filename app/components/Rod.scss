.probe {
  position: absolute;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
  z-index: 1;
  overflow: hidden;

  b {
    $radius: 25px;

    position: absolute;
    display: block;
    width: 2 * $radius;
    height: $radius;
    border-bottom-left-radius: $radius;
    border-bottom-right-radius: $radius;
    left: 50%;
    transform: translateX(-50%);
    background: radial-gradient(
      circle at top,
      #ccc,
      #ccc $radius * 0.25,
      #fff $radius * 0.25,
      #fff $radius * 0.75,
      rgba(255, 255, 255, 0.2) $radius * 0.8,
      rgba(255, 255, 255, 0.2) $radius
    );
    z-index: 2;
  }

  canvas {
    display: block;
    transform-origin: top;
    z-index: 1;
    transition: transform 1s linear;

    &.on {
      transform: rotate(30deg);
    }

    &.off {
      transform: rotate(0deg);
    }
  }
}
