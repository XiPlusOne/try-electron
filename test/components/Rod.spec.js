import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Rod from '../../app/components/Rod';

Enzyme.configure({ adapter: new Adapter() });

describe('Rod component', () => {
  const painter = {
    paint() {}
  };

  it('should match exact snapshot', () => {
    const rod = <Rod painter={painter} />;
    const tree = renderer.create(rod).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should painter should paint a rod', () => {
    const spyPainter = {
      paint: spy()
    };
    const component = mount(<Rod painter={spyPainter} />);

    expect(spyPainter.paint.called).toBe(true);
    component.unmount();
  });

  it('should canvas should has a class named on if rod is on', () => {
    const component = shallow(<Rod painter={painter} isRodOn />);

    const canvas = component.find('canvas').at(0);
    expect(canvas.hasClass('on')).toBe(true);
  });

  it('should canvas should has a class named off if rod is not on', () => {
    const component = shallow(<Rod painter={painter} isRodOn={false} />);

    const canvas = component.find('canvas').at(0);
    expect(canvas.hasClass('off')).toBe(true);
  });

  it('should canvas should call rodOnCallback if rod is on after transition is over', () => {
    const rodOnCallback = spy();
    const component = shallow(
      <Rod painter={painter} rodOnCallback={rodOnCallback} isRodOn />
    );

    const canvas = component.find('canvas').at(0);
    canvas.simulate('transitionend');
    expect(rodOnCallback.called).toBe(true);
  });

  it('should canvas should call rodOffCallback if rod is not on after transition is over', () => {
    const rodOffCallback = spy();
    const component = shallow(
      <Rod painter={painter} rodOffCallback={rodOffCallback} isRodOn={false} />
    );

    const canvas = component.find('canvas').at(0);
    canvas.simulate('transitionend');
    expect(rodOffCallback.called).toBe(true);
  });
});
