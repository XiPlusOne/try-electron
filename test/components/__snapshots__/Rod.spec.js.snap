// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rod component should match exact snapshot 1`] = `
<div
  className="probe"
  style={null}
>
  <b />
  <canvas
    className="off"
    onTransitionEnd={[Function]}
  />
</div>
`;
